### ISSUE TITLE
LSTM double backprop raise error

### ISSUE THREAD

**Issue by Hiroshiba** [chainer/chainer#4017]
I tried double backprop with LSTM. Then two error occurred.

```py3
import chainer
import numpy

print('chainer version:', chainer.__version__)

lstm = chainer.links.StatelessLSTM(3, 3)
x = chainer.Variable(numpy.random.rand(2, 3).astype(numpy.float32))
_, y = lstm(None, None, x)

g, = chainer.grad([y], [x], enable_double_backprop=True)
l = chainer.functions.mean_squared_error(g, numpy.ones_like(g.data, numpy.float32))
l.backward()
print(l)
```

First error,
```
chainer version: 3.1.0
Traceback (most recent call last):
  File "check.py", line 10, in <module>
    g, = chainer.grad([y], [x], enable_double_backprop=True)
  File "/hoge/lib/python3.6/site-packages/chainer/function_node.py", line 758, in grad
    _backprop(outputs, inputs, grad_required, retain_grad, grads)
  File "/hoge/lib/python3.6/site-packages/chainer/function_node.py", line 817, in _backprop
    new_gxs = func.backward_accumulate(input_indexes, gys, gxs)
  File "/hoge/lib/python3.6/site-packages/chainer/function_node.py", line 509, in backward_accumulate
    gxs = self.backward(target_input_indexes, grad_outputs)
  File "/hoge/lib/python3.6/site-packages/chainer/functions/activation/lstm.py", line 113, in backward
    self.get_retained_inputs() + self.get_retained_outputs() + grads)
TypeError: can only concatenate tuple (not "list") to tuple
```

This error may be caused that `grads` is not tuple.
So I changed such as `tuple(grads)`.
Then second error is occurred.

Second error,
```
chainer version: 3.1.0
Traceback (most recent call last):
  File "check.py", line 13, in <module>
    l.backward()
  File "/hoge/lib/python3.6/site-packages/chainer/variable.py", line 852, in backward
    self._backward_main(retain_grad)
  File "/hoge/lib/python3.6/site-packages/chainer/variable.py", line 953, in _backward_main
    target_input_indexes, out_grad, in_grad)
  File "/hoge/lib/python3.6/site-packages/chainer/function_node.py", line 509, in backward_accumulate
    gxs = self.backward(target_input_indexes, grad_outputs)
  File "/hoge/lib/python3.6/site-packages/chainer/function.py", line 139, in backward
    gxs = self._function.backward(in_data, grad_out_data)
  File "/hoge/lib/python3.6/site-packages/chainer/functions/activation/lstm.py", line 186, in backward
    ggc = ggc_prev.copy()
AttributeError: 'NoneType' object has no attribute 'copy'
```

I cannot understand why this error was occurred.

**Comment by murawaki** [comment]
This is just to say I need this bug to be fixed :-)

Looks like some weakref-ed VariableNode (and Variable?) objects are dead.

**Comment by unnonouno** [comment]
This problem is fixed via #4320. Your code works well now. If you face the same problem, please reopen the issue.
